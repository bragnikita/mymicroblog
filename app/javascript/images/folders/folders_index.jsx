import React, {Component} from 'react';
import PropTypes from 'prop-types';
import classes from 'classnames';
import _ from 'lodash';
import styles from './folders-styles.scss';

class AddRenameFolderForm extends Component {

    state = {
        name: this.props.name || '',
    };

    onChange = (e) => {
        this.setState({
            name: e.target.value,
        });
    };

    onApply = () => {
        this.props.onApply && this.props.onApply(this.state.name);
    };

    onCancel = () => {
        this.props.onCancel && this.props.onCancel();
    };

    isValidName = () => {
        return !!this.state.name;
    };

    render() {
        return (
            <div className={classes("w-100 d-flex align-items-center", this.props.className)}>
                <input name="input_folder_name"
                    className="form-control form-control-narrow filler mr-2" type="text" value={this.state.name}
                       onChange={this.onChange}/>
                <button
                    name="apply"
                    className="btn btn-success btn-circle mr-1"
                    onClick={this.onApply}
                    disabled={!this.isValidName()}
                >
                    <i className="fa fa-chevron-circle-down"/>
                </button>
                <button
                    name="cancel"
                    className="btn btn-warning btn-circle"
                    onClick={this.onCancel}
                >
                    <i className="fa fa-times"/>
                </button>
            </div>
        )
    }
}

class AddRenameFolderLine extends Component {

    static propTypes = {
      createNew: PropTypes.func,
      rename: PropTypes.func,
      name: PropTypes.string,
    };

    state = {
        mode: 'none',
    };

    reset = () => {
        this.setState({mode: 'none'});
    };

    createNew = (name) => {
        if (this.props.createNew) {
            this.props.createNew(name).then(() => this.reset())
        } else {
            this.reset();
        }
    };

    rename = (name) => {
        if (this.props.rename) {
            this.props.rename(name).then(() => this.reset())
        } else {
            this.reset();
        }
    };

    render() {
        if (this.state.mode === 'none') {
            return (
                <div className="w-100 d-flex justify-content-end">
                    <button
                        name="rename"
                        className="btn btn-outline-secondary btn-circle mr-1"
                        onClick={() => this.setState({mode: 'rename'})}
                        disabled={!this.props.name}
                    >
                        <i className="fa fa-i-cursor"/>
                    </button>
                    <button
                        name="add"
                        className="btn btn-success btn-circle"
                        onClick={() => this.setState({mode: 'add'})}
                    >
                        <i className="fa fa-plus"/>
                    </button>
                </div>
            );
        } else if (this.state.mode === 'add') {
            const props = {
                onApply: this.createNew,
                onCancel: this.reset,
            };

            return (<AddRenameFolderForm  {...props} />)
        } else {
            const props = {
                onApply: this.rename,
                onCancel: this.reset,
                name: this.props.name,
            };

            return (<AddRenameFolderForm  {...props} />)
        }

    }
}

export default class FoldersIndex extends Component {

    static propTypes = {
        handleSelect: PropTypes.func,
        requestFolders: PropTypes.func.isRequired,
        rename: PropTypes.func,
        createNew: PropTypes.func,
        initFolderId: PropTypes.string,
    };

    state = {
        folders: [],
        mode: null,
        current_id: this.props.initFolderId,
    };

    static defaultProps = {
        handleSelect: (newFolder) => {}
    };

    handleSelect = _.memoize((folder) => () => {
        this.setState({current_id: folder.id});
        this.props.handleSelect(folder)
    }, (folder) => folder.id);

    createItems = () => {
        return this.state.folders.map((f) => {
            const classNames = classes(
                "item btn-outline-secondary col-auto",
                {[styles.selected]: f.id === this.state.current_id}
            );
            return (
                <button key={f.id} className={classNames} onClick={this.handleSelect(f)} name="folder">
                    <i className="fa fa-folder-o"/> {f.name}
                </button>
            )
        });
    };

    currentFolderName = () => {
        const currentFolder = _.find(this.state.folders, {id: this.state.current_id});
        return currentFolder ? currentFolder.name : null;
    };

    createNew = (name) => {
        if (this.props.createNew) {
            return this.props.createNew(name).then(this.updateList)
        } else {
            return Promise.resolve()
        }
    };

    rename = (name) => {
        if (this.props.rename) {
            return this.props.rename(this.state.current_id, name).then(this.updateList)
        } else {
            return Promise.resolve()
        }
    };

    render() {

        const items = this.createItems();

        return (
            <div className={classes("folders-index w-100")}>
                <div className="container-fluid">
                    <div className="row">
                        {items}
                    </div>
                    <div className="row mt-2">
                        <div className="col-xs-12 col-lg-6 no-h-padding border-bottom p-1">
                            <AddRenameFolderLine
                                name={this.currentFolderName()}
                                createNew={this.createNew}
                                rename={this.rename}
                            />
                        </div>
                    </div>
                </div>
            </div>
        )
    }

    updateList = () => {
        this.props.requestFolders().then((folders) => {
            let currentFolder = _.find(folders, {id: this.state.current_id});
            if (!currentFolder) {
                currentFolder = _.first(folders);
                currentFolder && this.props.handleSelect(currentFolder);
            }
            this.setState({
                folders: folders,
                current_id: currentFolder ? currentFolder.id : null,
            })
        });
    };

    componentDidUpdate = (prevProps, prevState) => {

    };

    componentDidMount = () => {
        this.updateList();
    }
}

export {
    AddRenameFolderForm,
    AddRenameFolderLine
}