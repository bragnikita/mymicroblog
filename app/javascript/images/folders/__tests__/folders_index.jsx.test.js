import React from 'react';
import {shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import {AddRenameFolderForm, AddRenameFolderLine} from '../folders_index';


describe('AddRenameFolderForm', () => {
    describe('Snapshots', () => {
        let props = {};
        const baseProps = {};

        const subject = () => shallow(
            <AddRenameFolderForm {...props} />
        );

        beforeEach(() => {
            props = Object.assign({}, baseProps)
        });

        test('{name: "common"}', () => {
            props.name = 'common';
            expect(shallowToJson(subject())).toMatchSnapshot();
        });
        test('{name: undefined}', () => {
            expect(shallowToJson(subject())).toMatchSnapshot();
        });
        test('after some text inputted', () => {
            const component = subject();
            component.instance().onChange({target: {value: 'some'}});
            expect(shallowToJson(component)).toMatchSnapshot();
        });
    });
});

describe('AddRenameFolderLine', () => {
    let props = {};

    beforeEach(() => { props = {}});

    const subject = (extraProps = {}) => shallow(
        <AddRenameFolderLine {...props} {...extraProps} />
    );
    describe('Snapshots', () => {

        test('if folder name specified', () => {
            props = {name: 'folder_name'};
            expect(shallowToJson(subject())).toMatchSnapshot();
        });
        test('if folder name is not specified', () => {
            props = {};
            expect(shallowToJson(subject())).toMatchSnapshot();
        });
        test('when new button was clicked', () => {
            const component = subject();
            component.find('[name="add"]').first().simulate('click');
            expect(shallowToJson(component)).toMatchSnapshot();
        });
        test('when rename button was clicked', () => {
            const component = subject({name: 'new folder'});
            component.find('[name="rename"]').first().simulate('click');
            expect(shallowToJson(component)).toMatchSnapshot();
        });
    });
    describe('Callbacks', () => {
        test('it calls rename', () => {
            const rename = jest.fn().mockReturnValue(Promise.resolve());
            const component = subject({rename, name: "old_name"});
            component.find('[name="rename"]').first().simulate('click');
            component.find(AddRenameFolderForm).prop('onApply')('renamed_folder');
            expect(rename.mock.calls.length).toBe(1);
            expect(rename.mock.calls[0][0]).toBe('renamed_folder');
        });
        test('it calls createNew', () => {
            const createNew = jest.fn().mockReturnValue(Promise.resolve());
            const component = subject({createNew});
            component.find('[name="add"]').first().simulate('click');
            component.find(AddRenameFolderForm).prop('onApply')('new_folder');
            expect(createNew.mock.calls.length).toBe(1);
            expect(createNew.mock.calls[0][0]).toBe('new_folder');
        })
    })
});