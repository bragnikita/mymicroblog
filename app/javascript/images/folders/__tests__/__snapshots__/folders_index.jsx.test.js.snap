// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRenameFolderForm Snapshots {name: "common"} 1`] = `
<div
  className="w-100 d-flex align-items-center"
>
  <input
    className="form-control form-control-narrow filler mr-2"
    name="input_folder_name"
    onChange={[Function]}
    type="text"
    value="common"
  />
  <button
    className="btn btn-success btn-circle mr-1"
    disabled={false}
    name="apply"
    onClick={[Function]}
  >
    <i
      className="fa fa-chevron-circle-down"
    />
  </button>
  <button
    className="btn btn-warning btn-circle"
    name="cancel"
    onClick={[Function]}
  >
    <i
      className="fa fa-times"
    />
  </button>
</div>
`;

exports[`AddRenameFolderForm Snapshots {name: undefined} 1`] = `
<div
  className="w-100 d-flex align-items-center"
>
  <input
    className="form-control form-control-narrow filler mr-2"
    name="input_folder_name"
    onChange={[Function]}
    type="text"
    value=""
  />
  <button
    className="btn btn-success btn-circle mr-1"
    disabled={true}
    name="apply"
    onClick={[Function]}
  >
    <i
      className="fa fa-chevron-circle-down"
    />
  </button>
  <button
    className="btn btn-warning btn-circle"
    name="cancel"
    onClick={[Function]}
  >
    <i
      className="fa fa-times"
    />
  </button>
</div>
`;

exports[`AddRenameFolderForm Snapshots after some text inputted 1`] = `
<div
  className="w-100 d-flex align-items-center"
>
  <input
    className="form-control form-control-narrow filler mr-2"
    name="input_folder_name"
    onChange={[Function]}
    type="text"
    value="some"
  />
  <button
    className="btn btn-success btn-circle mr-1"
    disabled={false}
    name="apply"
    onClick={[Function]}
  >
    <i
      className="fa fa-chevron-circle-down"
    />
  </button>
  <button
    className="btn btn-warning btn-circle"
    name="cancel"
    onClick={[Function]}
  >
    <i
      className="fa fa-times"
    />
  </button>
</div>
`;

exports[`AddRenameFolderLine Snapshots if folder name is not specified 1`] = `
<div
  className="w-100 d-flex justify-content-end"
>
  <button
    className="btn btn-outline-secondary btn-circle mr-1"
    disabled={true}
    name="rename"
    onClick={[Function]}
  >
    <i
      className="fa fa-i-cursor"
    />
  </button>
  <button
    className="btn btn-success btn-circle"
    name="add"
    onClick={[Function]}
  >
    <i
      className="fa fa-plus"
    />
  </button>
</div>
`;

exports[`AddRenameFolderLine Snapshots if folder name specified 1`] = `
<div
  className="w-100 d-flex justify-content-end"
>
  <button
    className="btn btn-outline-secondary btn-circle mr-1"
    disabled={false}
    name="rename"
    onClick={[Function]}
  >
    <i
      className="fa fa-i-cursor"
    />
  </button>
  <button
    className="btn btn-success btn-circle"
    name="add"
    onClick={[Function]}
  >
    <i
      className="fa fa-plus"
    />
  </button>
</div>
`;

exports[`AddRenameFolderLine Snapshots when new button was clicked 1`] = `
<AddRenameFolderForm
  onApply={[Function]}
  onCancel={[Function]}
/>
`;

exports[`AddRenameFolderLine Snapshots when rename button was clicked 1`] = `
<AddRenameFolderForm
  name="new folder"
  onApply={[Function]}
  onCancel={[Function]}
/>
`;
